[새기능] 합격 처리시 타지역 출신 합격자들을 최대 50%로 제한

jyj1289 opened the pull request:
## 🎫 관련 이슈
[//]: # (다음 키워드를 사용하면 해당 PR을 머지할 때 자동으로 이슈를 닫을 수 있습니다.)
[//]: # (keyword: close|closes|closed|resolve|resolves|resolved|fix|fixes|fixed)
[//]: # (예시: close #1)

close #200 

<br>

## 📄 개요
[//]: # (작업 내용을 간단히 요약해서 적습니다.)
[//]: # (예시: 유저 회원가입 기능을 만들었습니다.)

> 타지역 출신 지원자의 선발 비율이 전체 모집 인원의 50%를 넘지 않도록 제한하는 기능을 구현했습니다.

<br>

## 🔨 작업 내용
[//]: # (작업 내용을 자세하게 적습니다.)
[//]: # (붙임표 "-" 을 사용해서 목록을 만듭니다.)
[//]: # (예시: 유저 회원가입 API를 만들었습니다.)

- SelectFirstFormUseCase와 SelectSecondFormUseCase에 제한하는 기능을 구현했습니다.
- SelectFirstFormUseCaseTest와 SelectSecondFormUseCaseTest를 수정하였습니다.
- FormFixture에 generateFormList와 generateRandomForm을 각각 generateBusanFormList, genearteOtherRegionFormList, generateRandomBusanForm, generateRandomOtherRegionForm으로 분리했습니다.

<br>

## 🏁 확인 사항
[//]: # (PR을 보내기 전 다음 사항을 확인해주세요.)
[//]: # (해당 사항을 모두 이행해야 머지할 수 있습니다.)
[//]: # (- [x] 를 사용해서 완료로 표시할 수 있습니다.)

- [x] 테스트를 완료했나요?
- [ ] API 문서를 작성했나요?
- [x] 코드 컨벤션을 준수했나요?
- [x] 불필요한 로그, 주석, import 등을 삭제했나요?

<br>

## 🙋🏻 덧붙일 말
[//]: # (다음 사항이 있다면 적어주세요.)
[//]: # (PR에 대한 추가 설명)
[//]: # (중점적으로 리뷰받고 싶은 부분)
[//]: # (기타 등등)

jyj1289 committed:
feat(#200): 원서 1차 및 2차 합격 조건 추가
- 타지역 지원자들은 총원의 50%를 초과할 수 없도록 조건을 추가했어요.

jyj1289 committed:
test(#200): generateBusanFormList() 및 generateOtherRegionFormList() 메서드 추가
- generateRandomForm() 메서드를 generateRandomBusanForm()과 generateRandomOtherRegionForm() 메서드로 분리했어요.
- generateFormList()를 generateBusanFormList()와 generateOtherRegionFormList() 메서드로 분리했어요.

jyj1289 committed:
test(#200): 타지역 50% 이하 선발 검사
- 타지역 지원자는 총원의 50% 이하로 선발되어야해서 검사 조건을 추가했어요.

gimhanul commented on src/main/java/com/bamdoliro/maru/application/form/SelectSecondPassUseCase.java:
```diff
@@ -20,6 +21,7 @@ public class SelectSecondPassUseCase {
 
     private final FormRepository formRepository;
     private final CalculateFormScoreService calculateFormScoreService;
+    private int otherRegionCount = FixedNumber.TOTAL / 2;
```
이런 설정값들 설정값 상수 폴더에 몰아 넣어져있지 않나용?

jyj1289 replied:
FixedNumber에 넣는걸로 변경했습니다.

gimhanul commented on src/main/java/com/bamdoliro/maru/application/form/SelectSecondPassUseCase.java:
```diff
@@ -73,8 +75,16 @@ public void execute() {
+                    count--;
+                } else if (otherRegionCount > 0) {
+                    form.pass();
+                    otherRegionCount--;
```
멀티 스레드 환경에서 전역 변수를 쓰면 동시성 문제가 있을 거 같구
(동시에 쏠 거 같지는 않지만...)

아무튼 그래서 지역 변수로 쓰는 게 좋을 거 같고
혹시 모를 상황을 대비해서 이 메서드를 2번 이상 호출했을 때에 대한 처리도 필요해보입니다.

jyj1289 replied:
동시성 문제를 해결하기 위해, 해당 변수를 AtomicInteger 타입의 지역 변수로 변경하였습니다.

2번 이상 호출을 하더라도 이미 실행이 완료되어 1차합격자가 없어지기 때문에 다시 실행해도 문제가 없을 것 같습니다. 
그래도 만약을 대비해 처리를 하는 것이 좋을까요?

gimhanul replied:
지역 변수) Atomic Integer는 지역 변수로 사용할 때는 필요 없다고 생각됩니다. 메서드 내에서 변수 선언 시 하나의 스레드에서만 접근 할 수 있으니까용 Atomic으로 유지하고 싶은 거면 Atomic + 전역 변수 조합으로 사용하거나 primitive + 지역 조합으로 쓰는 게 맞아요
메서드 두 번 호출 방지) 높은 우선순위는 아닌 거 같고, 나중에 명시적으로 에러를 던져주면 좋을 거 같아요

jyj1289 replied:
Atomic integer를 전역 변수로 변경하였습니다.

메서드 호출 방지는 추후에 명시적으로 에러를 발생시키도록 하겠습니다. 
감사합니다 :)

gimhanul reviewed (approved):
뭔가 2025년의 '마루' 서비스의 터닝 포인트가 되는 연도같아용
앞으로의 방향성을 잘 결정해서 잡아나가길...

저는 설정값들을 어드민이 매년 커스텀하게 스스로 바꿀 수 있도록 하는 기능 같은 걸 만들면 좋을 거 같아요 (기간 설정, 부산/타지역 비율 설정 등)

cabbage16 commented on src/test/java/com/bamdoliro/maru/application/form/SelectSecondPassUseCaseTest.java:
```diff
@@ -56,8 +57,8 @@ void setUp() {
-        List<Form> formList = FormFixture.generateFormList(userList);
-        formList.forEach(form -> {
+        List<Form> formList = FormFixture.generateBusanFormList(userList.subList(0, userList.size() / 2));
+        formList.addAll(FormFixture.generateOtherRegionFormList(userList.subList(userList.size() / 2, userList.size())));        formList.forEach(form -> {
```
P4) 이쪽에 줄바꿈이 빠진 거 같습니다.

jyj1289 replied:
줄바꿈 적용했습니다.

jyj1289 committed:
feat(#200): 부산/타지역 비율 상수화
- 타지역 비율을 임의로 넣지 않게하기위해서 FixedNumber에 OTHER_REGION_RATE를 추가했어요.

jyj1289 committed:
fix(#200): otherRegionCount 지역변수로 변경
- 동시성 문제를 해결하기 위해서 otherRegionCount를 지역변수로 변경했어요.
- 총원의 50% 만큼 선발하기 위해서는 변수를 참조해야하기때문에 AtomicInteger를 사용했어요.

jyj1289 committed:
test(#200): 줄바꿈
- 가독성을 위해 줄을 변경했어요.

jyj1289 committed:
fix(#200): FixedNumber 값 복구
- 테스트를 위해 변경했던 값을 원래 값으로 복구했어요.

jyj1289 committed:
refactor(#200): otherRegionCount 전역변수로 전환
- AtomicInteger를 사용하면 지역변수가 아닌 전역변수로 변수를 선언하는게 맞다고 판단되어 수정하였습니다.
